File: src/Matriz/CGNR.java
package Matriz;

import org.jblas.DoubleMatrix;

public class CGNR {
    
    public int iteractions = 0;
    public DoubleMatrix getImage(DoubleMatrix H, DoubleMatrix g, Integer selectedModelFlag) {
        System.out.println("H rows: " + H.rows + " x " + H.columns + " and g rows: " + g.columns);

        DoubleMatrix f = DoubleMatrix.zeros(H.columns);
        DoubleMatrix r = g.sub(H.mmul(f));

        DoubleMatrix z = H.transpose().mmul(r);
        DoubleMatrix p = z;
        int i = 0;

        do{
            DoubleMatrix w = H.mmul(p);
            Double z_squared = z.norm2() * z.norm2();
            Double w_squared = w.norm2() * w.norm2();

            Double alpha = z_squared / w_squared;
            
            DoubleMatrix f1 = f.add(p.mul(alpha));
            DoubleMatrix r1 = r.sub(w.mul(alpha));
            DoubleMatrix z1 = H.transpose().mmul(r1);

            Double z1_squared = z1.norm2() * z1.norm2();
            Double beta = z1_squared / z_squared;
            DoubleMatrix p1 = z1.add(p.mul(beta));

            f = f1;
            r = r1;
            z = z1;
            p = p1;
            i++;

        }while(i < 30);

        if (selectedModelFlag == 1) {
            f.reshape(60, 60);
        }
        else {
            f.reshape(30, 30);
        }

        return f;
    }
}

File: CGNR.py
import org.jblas.DoubleMatrix

class CGNR:
    def __init__(self):
        self.iterations = 0

    def get_image(self, h, g, selected_model_flag):
        f = DoubleMatrix.zeros(h.columns)
        r = g - h.mmul(f)
        z = h.transpose().mmul(r)
        p = z

        error = float('inf')
        i = 0
        while error > 1e-4 and i < 30:
            w = h.mmul(p)
            z_squared = z.norm2() ** 2
            w_squared = w.norm2() ** 2
            alpha = z_squared / w_squared

            new_f = f + p * alpha
            new_r = r - w * alpha
            new_z = h.transpose().mmul(new_r)
            new_z_squared = new_z.norm2() ** 2
            beta = new_z_squared / z_squared
            new_p = new_z + p * beta

            error = self.error(new_r, r)

            f = new_f
            r = new_r
            z = new_z
            p = new_p
            i += 1

        self.iterations = i

        if selected_model_flag == 1:
            f.reshape(60, 60)
        else:
            f.reshape(30, 30)

        return f

    def get_iterations(self):
        return self.iterations

    def error(self, new_r, r):
        return (new_r - r).norm2()